File: book-talk-be/_lib/src/main/kotlin/kr/co/booktalk/client/MonitorClient.kt
package kr.co.booktalk.client

import io.github.oshai.kotlinlogging.KotlinLogging
import io.ktor.client.*
import io.ktor.client.request.*
import io.ktor.http.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kr.co.booktalk.config.LibProperties

/**
 * 모니터링 클라이언트
 */
interface MonitorClient {
    suspend fun send(request: SendRequest): Unit
}

data class SendRequest(
    val title: String,
    val message: String,
    val stackTrace: String,
    val level: Level,
) {
    enum class Level {
        ERROR
    }
}

/**
 * see [Slack API](https://api.slack.com/messaging/webhooks)
 */
class SlackMonitorClient(
    private val httpClient: HttpClient,
    private val properties: LibProperties.SlackProperties,
) : MonitorClient {
    private val logger = KotlinLogging.logger {}

    override suspend fun send(request: SendRequest) {
        runCatching {
            val response = withContext(Dispatchers.IO) {
                httpClient.post(properties.webhookUrl) {
                    contentType(ContentType.Application.Json)
                    setBody(mapOf("blocks" to buildBlocks(request)))
                }
            }

            if (!response.status.isSuccess()) {
                logger.error { "Slack 알림 전송 실패: ${response.status}" }
            }
        }.onFailure {
            /** livelock 방지를 위해 예외 전파 x */
            logger.error(it) { "Slack 알림 처리 중 예외 발생" }
        }
    }

    private fun buildBlocks(request: SendRequest): List<Any> {
        return when (request.level) {
            SendRequest.Level.ERROR -> listOf(
                headerBlock(":rotating_light: Server Error"),
                sectionBlock(
                    """
                *- message*
                ${request.message}
                """.trimIndent()
                ),
                dividerBlock(),
                sectionBlock(
                    """
                *- stack trace*
                ``` 
                ${request.stackTrace}
                ```
                """.trimIndent()
                )
            )
        }
    }

    private fun headerBlock(text: String) =
        mapOf(
            "type" to "header",
            "text" to mapOf(
                "type" to "plain_text",
                "text" to text
            )
        )

    private fun sectionBlock(text: String) =
        mapOf(
            "type" to "section",
            "text" to mapOf(
                "type" to "mrkdwn",
                "text" to text
            )
        )

    private fun dividerBlock() =
        mapOf("type" to "divider")
}

class NoOpMonitorClient : MonitorClient {
    private val logger = KotlinLogging.logger {}

    override suspend fun send(request: SendRequest) {
        logger.warn { "[NoOp] monitor send: $request" }
    }
}

File: book-talk-ts/packages/ai/src/client/monitor.client.ts
import { Logger } from '@nestjs/common';
import axios from 'axios';

export const MONITOR_CLIENT = Symbol.for('MONITOR_CLIENT');

/** 알림 요청 */
export interface SendRequest {
  title: string;
  message: string;
  stackTrace: string;
  level: 'ERROR';
}

/** 모니터링 클라이언트 인터페이스 */
export interface MonitorClient {
  send(request: SendRequest): Promise<void>;
}

/** Slack 모니터링 클라이언트 */
export class SlackMonitorClient implements MonitorClient {
  constructor(private readonly webhookUrl: string) {}

  async send(request: SendRequest): Promise<void> {
    try {
      const response = await axios.post(this.webhookUrl, {
        blocks: this.buildBlocks(request),
      });

      if (response.status < 200 || response.status >= 300) {
        Logger.error(`Slack 알림 전송 실패: ${response.status}`);
      }
    } catch (error) {
      /** livelock 방지를 위해 예외 전파 x */
      Logger.error(`Slack 알림 처리 중 예외 발생: ${error}`);
    }
  }

  private buildBlocks(request: SendRequest): object[] {
    switch (request.level) {
      case 'ERROR':
        return [
          this.headerBlock(`:rotating_light: ${request.title}`),
          this.sectionBlock(`*- message*: ${request.message}`),
          this.sectionBlock(`*- stack trace*:\n\`\`\`${request.stackTrace}\`\`\``),
        ];
    }
  }

  private headerBlock(text: string) {
    return {
      type: 'header',
      text: { type: 'plain_text', text },
    };
  }

  private sectionBlock(text: string) {
    return {
      type: 'section',
      text: { type: 'mrkdwn', text },
    };
  }
}

/** NoOp 모니터링 클라이언트 */
export class NoOpMonitorClient implements MonitorClient {
  async send(request: SendRequest): Promise<void> {
    Logger.warn(`[NoOp] monitor send: ${JSON.stringify(request)}`);
  }
}
